odoo/addons/base/tests/test_misc.py

```diff
@@ -468,3 +477,26 @@ def test_readonly_dict(self):
             d.update({'baz': 'xyz'})
         with self.assertRaises(TypeError):
             dict.update(d, {'baz': 'xyz'})
+
+
+class TestUrlValidate(BaseCase):
+    def test_url_validate(self):
+        for case, truth in [
+            # full URLs should be preserved
+            ('http://example.com', 'http://example.com'),
+            ('http://example.com/index.html', 'http://example.com/index.html'),
+            ('http://example.com?debug=1', 'http://example.com?debug=1'),
+            ('http://example.com#h3', 'http://example.com#h3'),
+
+            # URLs with a domain should get a http scheme
+            ('example.com', 'http://example.com'),
+            ('example.com/index.html', 'http://example.com/index.html'),
+            ('example.com?debug=1', 'http://example.com?debug=1'),
+            ('example.com#h3', 'http://example.com#h3'),
+        ]:
+            self.assertEqual(validate_url(case), truth)
+
+        # broken cases, do we really want that?
+        self.assertEqual(validate_url('/index.html'), 'http:///index.html')
```

Comment:
note that `.zip` and such exist now, so not ideal if someone has a relative url to `/hackme.zip`. Not that it's very likely to have a url to a file directly at the base path but who knows.

Reply:
Since we are mentioning it, here is a complete blog post about the .zip TLD https://www.fortinet.com/blog/industry-trends/threat-actors-add-zip-domains-to-phishing-arsenals

https://odoo.com∕web∕content&id=172&filename=@chatgpt5.zip
```py
from urllib.parse import urlsplit
#               U+2215 division slash
#                     |   |
url = "http://odoo.com∕web∕content&id=172&filename=@chatgpt5.zip"
#      ^^^^   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#     scheme                         netloc
#             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^
#                           username                  hostname

urlsplit(url).hostname # chatgpt5.zip
```
```bash
$ httpx -v "http://odoo.com∕web∕content&id=172&filename=@chatgpt5.zip"
* Connecting to 'chatgpt5.zip'
* Connected to '76.76.21.9' on port 80
GET / HTTP/1.1
Host: chatgpt5.zip
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
User-Agent: python-httpx/0.26.0
Authorization: Basic b2Rvby5jb23iiJV3ZWLiiJVjb250ZW50JmlkPTE3MiZmaWxlbmFtZT06

HTTP/1.0 308 Permanent Redirect
Content-Type: text/plain
Location: https://chatgpt5.zip/
Refresh: 0;url=https://chatgpt5.zip/
server: Vercel

$ # the "Authorization: Basic" is just the base64 of f"{username}:{password}"
$ echo b2Rvby5jb23iiJV3ZWLiiJVjb250ZW50JmlkPTE3MiZmaWxlbmFtZT06 | base64 -d
odoo.com∕web∕content&id=172&filename=:
```

Reply:
Truly cursed